### PR TITLE
Throw MemberAccessException when abstract class is used with newobj instruction

### PR DESCRIPTION
**Pull request by joncham**
Previously an abstract instance would be created and potentially crash when vtable members were accessed.

### CODE REVIEW

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -3641,6 +3641,14 @@ handle_alloc (MonoCompile *cfg, MonoClass *klass, gboolean for_box, int context_
 	MonoInst *iargs [2];
 	void *alloc_ftn;
 
+	if (mono_class_get_flags (klass) & TYPE_ATTRIBUTE_ABSTRACT)	{
+		char* full_name = mono_type_get_full_name (klass);
+		cfg->exception_message = g_strdup_printf ("Cannot create an abstract class: %s", full_name);
+		g_free (full_name);
+		mono_cfg_set_exception (cfg, MONO_EXCEPTION_MEMBER_ACCESS);
```
Comment (lambdageek) [inline mono/mini/method-to-ir.c]: Could do
```c
		mono_cfg_set_exception (cfg, MONO_EXCEPTION_MONO_ERROR);
		mono_error_set_generic_error (&cfg->error, "System", "MethodAccessException", "Cannot crerate an abstract class: %s", full_name);
		g_free (full_name);
```

For bonus points add `mono_error_set_member_access (MonoError *error, const char *msg_format, ...) MONO_ATTR_FORMAT_PRINTF(2,3)` to `utils/mono-error-internals.h` analogously to `mono_error_set_invalid_program` and others.

File: mono/utils/mono-error.c [hunk 2]
```diff
@@ -588,6 +599,9 @@ mono_error_prepare_exception (MonoError *oerror, MonoError *error_out)
 	case MONO_ERROR_MISSING_FIELD:
 		exception = mono_corlib_exception_new_with_args ("System", "MissingFieldException", error->full_message, error->first_argument, error_out);
 		break;
+	case MONO_ERROR_MEMBER_ACCESS:
+		exception = mono_corlib_exception_new_with_args ("System", "MemberAccessException", error->full_message, error->first_argument, error_out);
```
Comment (lambdageek) [inline mono/utils/mono-error.c]: I think this should be `mono_exception_from_name_msg (mono_defaults.corlib, "System", "MemberAccessException", error->full_message)`.  There's no `MemberAccessException(String,String)` constructor, afaict.

File: mono/mini/method-to-ir.c [hunk 3]
```diff
@@ -3641,6 +3641,14 @@ handle_alloc (MonoCompile *cfg, MonoClass *klass, gboolean for_box, int context_
 	MonoInst *iargs [2];
 	void *alloc_ftn;
 
+	if (mono_class_get_flags (klass) & TYPE_ATTRIBUTE_ABSTRACT) {
+		char* full_name = mono_type_get_full_name (klass);
+		mono_cfg_set_exception (cfg, MONO_EXCEPTION_MONO_ERROR);
+		mono_error_set_member_access (&cfg->error, "Cannot crerate an abstract class: %s", full_name);
```
Comment (Therzok) [inline mono/mini/method-to-ir.c]: typo: create

**Comment by joncham** [comment]
@lambdageek I notice my test fails in AOT mode since the AOT compile fails and it's thus attempted to be JIT'd later on: https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-fullaot/9506/testReport/junit/MonoTests/runtime/newobj_abstract_exe/

**Comment by lambdageek** [comment]
@joncham That makes sense. AOT turned it effectively into a compile-time error.

@vargaz @alexanderkyte Do we have a way of writing negative tests for AOT?

@joncham if we can't turn it into a negative test, just turn the test off for AOT - add it to somewhere around https://github.com/mono/mono/blob/master/mono/tests/Makefile.am#L1015

**Comment by lambdageek** [comment]
@lewurm @BrzVlad Looks like the interpreter needs to handle newobj of an abstract class, too.

**Comment by BrzVlad** [comment]
let's disable the test on interp

**Comment by lewurm** [comment]
I pushed a fix for the interpreter

**Comment by lambdageek** [comment]
I pushed to disable the test on FullAOT

**Comment by lambdageek** [comment]
@monojenkins rebase

**Comment by lambdageek** [comment]
@monojenkins build failed

**Pull request merged**